File: AoC2023/Day6/cpp/day6.cpp


#include <iostream>
#include <vector>

int CalcRace(int race_duration, int record_distance)
{
    int possibilities = 0;

    int min_speed_to_break_record = (record_distance / (race_duration - 1)) + 1;

    for (int speed = min_speed_to_break_record; speed < race_duration; ++speed)
    {
        if (speed * (race_duration - speed) > record_distance)
        {
            possibilities++;
        }
    }

    return possibilities;
}

int CalcAllRaces(const std::vector<int> &race_times, const std::vector<int> &race_records)
{
    int total = 1;

    for (size_t i = 0; i < race_times.size(); ++i)
    {
        total *= CalcRace(race_times[i], race_records[i]);
    }

    return total;
}

int main()
{
    std::vector<int> input_times = {56, 97, 78, 75};
    std::vector<int> input_dist = {546, 1927, 1131, 1139};

    std::cout << "Total possibilities: " << CalcAllRaces(input_times, input_dist) << std::endl;

    return 0;
}

File: AoC2023/Day6/python/day6.py
def calc_race(race_duration, record_distance):
    possibilities = 0

    for speed in range(1, race_duration):
        if speed*(race_duration-speed) > record_distance:
            possibilities += 1
    return possibilities


input_times = [56, 97, 78, 75]
input_dist = [546, 1927, 1131, 1139]


def Calc_all_races(races_times, races_records):
    total = 1
    for time, record in zip(races_times, races_records):
        total *= calc_race(time, record)
    return total


print(Calc_all_races(input_times, input_dist))
